### PR TITLE
Improve errors messages thrown from a `transform` closure

### PR DESCRIPTION
**Pull request by rjstelling**
> This PR is an improved version after some feedback on #110. However, I wanted to to make sure this is still the right direction, and I'm happy for more feedback, especially because this PR is more substantial. 

## Rational

If an argument property supplies a `transform` closure, like so,

```swift
@Option(help: "Some string input please.", transform: { try convert($0) })
var input: String
```

...and if the function `convert(...)` throws, this error should be printed to standard error, the same way throwing from `run()` or `validate()` does.

I raised this issue on the [Swift Forums](https://forums.swift.org/t/throwing-from-a-transform-closure/35137) and it was suggested by @natecook1000 this was a bug.

### tl;dr

Improves default error messages when `transform` closures throw errors. Adds the ability to print custom error messages when `transform` closures throw. 

### Detail
1. Added a file called TransformEndToEndTests.swift that test simple cases of using `transform` closures with `@Option`and `@Argument` properties.
2. Added error message tests for custom and default errors.
3. In the `init(...)` calls of `Option` and `Argument` any errors thrown by the `transform` closure care caught and converted into an `ParserError.unableToParseValue`
4. Added a `customMessage` associated value to `ParserError.unableToParseValue`, this is a `String` optional. When this is set the default error message is suppressed and `customMessage` is used.
5. Added a public protocol called `CustomParserErrorConvertible`. Errors can conform to this to provide there own custom messages. Added conformance to `ValidationError`.

## Before merging

1. Discuss the usage of `CustomParserErrorConvertible`. The only other alternative I considered was using `localizedDescription`. But this would require another mechanism to use the default error message.
3. Update documentation 

### Checklist

- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: Sources/ArgumentParser/Parsing/ParserError.swift [hunk 1]
```diff
@@ -25,7 +25,7 @@ enum ParserError: Error {
   case duplicateExclusiveValues(previous: InputOrigin, duplicate: InputOrigin, originalInput: [String])
   /// We need a value for the given key, but it’s not there. Some non-optional option or argument is missing.
   case noValue(forKey: InputKey)
-  case unableToParseValue(InputOrigin, Name?, String, forKey: InputKey)
+  case unableToParseValue(InputOrigin, Name?, String, forKey: InputKey, customMessage: String?)
```
Comment (jayrhynas) [inline Sources/ArgumentParser/Parsing/ParserError.swift:28]: `customMessage` could have a default `nil` value so that you don't need to specify it every time

Comment (rjstelling) [reply Sources/ArgumentParser/Parsing/ParserError.swift]: Adding a default value to the `customMessage` parameter also reduced the file footprint of this PR.

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 2]
```diff
@@ -339,16 +339,18 @@ extension ErrorMessageGenerator {
...
-    case let (_, v?):
+    case let (_, v?, _):
       return "The value '\(value)' is invalid for '<\(v)>'"
-    case let (n?, _):
+    case let (n?, _, _):
       return "The value '\(value)' is invalid for '\(n.synopsisString)'"
-    case (nil, nil):
+    case (nil, nil, _):
       return "The value '\(value)' is invalid."
     }
```
Comment (jayrhynas) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:355]: I suppose it's a matter of preference, but I feel like we should just keep the context, and append the custom message if it is not `nil`. This would also remove the need for `CustomParserErrorConvertible` and instead just use `localizedDescription`.

Comment (rjstelling) [reply Sources/ArgumentParser/Usage/UsageGenerator.swift:355]: My preference is not to append custom messages to any default error or context. It's not really a custom message at that point, if I'm customising the error I want full control.

But I agree the context is useful. And using something like `localizedDescription` would make it cleaner. 

Maybe we should consider wrapping the original error (thrown from the transform closure) in the `unableToParseValue` instead of a custom error message. 

But then we'd need some error formatting option, possibly in `CommandConfiguration`?

Comment (natecook1000) [reply Sources/ArgumentParser/Usage/UsageGenerator.swift:355]: I agree with @jayrhynas here — the printed error for this should combine the existing error output with a string from the thrown error. The output for a transformation throws on non-alphanumeric input could look like this:

```
$ command --option bad-input-#!@#
Error: The value 'bad-input' is invalid for '--option': Contains invalid characters '#!@#'
Usage: command [--option <option>] ...
``` 

where `Contains invalid characters '#!@#'` is the string representation of the thrown error.

> Maybe we should consider wrapping the original error (thrown from the transform closure) in the unableToParseValue instead of a custom error message.

This sounds good! We don't want to add a protocol just for this functionality, so doing a best effort conversion of the error to a string would be great. In general we try to do that by trying to match on `LocalizedError where error.errorDescription != nil` and then just falling back to `String(describing: error)`. That makes it simple for people to have nice error messages by just conforming their error type to `CustomStringConvertible`.

Comment (rjstelling) [reply Sources/ArgumentParser/Usage/UsageGenerator.swift]: I have updated `unableToParseValue` to take an optional `Error` associated value. 

If this error is not nil `unableToParseValueMessage(…)` makes best-efforts to create a custom error message.

> The output for a transformation throws on non-alphanumeric input could look like this:
> 
> ```
> $ command --option bad-input-#!@#
> Error: The value 'bad-input' is invalid for '--option': Contains invalid characters '#!@#'
> Usage: command [--option <option>] ...
> ```
> 
> where Contains invalid characters '#!@#' is the string representation of the thrown error.

This is exactly how the errors are now formatted. "Contains invalid characters '#!@#'" must be returned by either `errorDescription` form a `LocalizedError` or `description` from `CustomStringConvertible`. If the error is `nil` the errors are formatted as before.

`ValidationError` works as expected, printing the `message` string.

Simple `enum`s, such as:

```swift
enum FooError: Error {
    case outOfBounds
    case unknown 
}
```

Simply print out the name of the `case`.

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 3]
```diff
@@ -339,16 +339,36 @@ extension ErrorMessageGenerator {
...
+            case is LocalizedError where (error as? LocalizedError)?.errorDescription != nil:
+                return (error as? LocalizedError)?.errorDescription
+            case let err?:
+                return String(describing: err)
+            default:
+                return nil
+        }
+    }()
+    
+    switch (name, valueName, customErrorMessage) {
```
Comment (natecook1000) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:360]: Since we should be doing the same thing in each case, this switch statement would be easier to understand if we just concatenate the `"The value ..."` text with the custom error message (if it exists).

File: Sources/ArgumentParser/Parsable Properties/Option.swift [hunk 4]
```diff
@@ -239,8 +239,16 @@ extension Option {
...
       var arg = ArgumentDefinition(kind: kind, help: help, parsingStrategy: ArgumentDefinition.ParsingStrategy(parsingStrategy), update: .unary({
-        (origin, _, valueString, parsedValues) in
-        parsedValues.set(try transform(valueString), forKey: key, inputOrigin: origin)
+        (origin, name, valueString, parsedValues) in
+        
+        do {
+            let transformedValue = try transform(valueString)
+            parsedValues.set(transformedValue, forKey: key, inputOrigin: origin)
+        }
+        catch /* All other errors */ {
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsable Properties/Option.swift:248]: Nit: don't start a new line before `catch`, here and elsewhere.

```suggestion
        } catch {
```

File: Sources/ArgumentParser/Parsable Properties/Errors.swift [hunk 5]
```diff
@@ -20,7 +20,7 @@ import MSVCRT
 /// An error type that is presented to the user as an error with parsing their
 /// command-line input.
 public struct ValidationError: Error, CustomStringConvertible {
-  var message: String
+  public internal(set) var message: String
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsable Properties/Errors.swift:23]: This is a good idea, but since it's an API change, let's handle it in a separate PR.

Comment (rjstelling) [reply Sources/ArgumentParser/Parsable Properties/Errors.swift:23]: This was a hangover from the previous protocol based design. I will open a separate PR.

File: Sources/ArgumentParser/Usage/UsageGenerator.swift (hunk 3 above)
Comment (rjstelling) [reply Sources/ArgumentParser/Usage/UsageGenerator.swift:360]: Trying to be a bit too clever and adding complexity in the process. 

I think some of these cases will never have custom messages, but it will just work if they ever do start passing in errors.

**Comment by rjstelling** [comment]
I have also updated the "Validation and Errors" file with a section on "Handling Transform Errors”.

Its not my greatest work 😬

Hopefully by the time anyone see it I will have had chance to have another draft or 2.

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 6]
```diff
@@ -343,33 +343,29 @@ extension ErrorMessageGenerator {
...
-                return (error as? LocalizedError)?.errorDescription
-            case let err?:
-                return String(describing: err)
-            default:
-                return nil
-        }
+    let customErrorMessage: String = {
+      switch error {
+      case is LocalizedError where (error as? LocalizedError)?.errorDescription != nil:
+        return ": " + (error as! LocalizedError).errorDescription! // !!! Checked above that this will not be nil
```
Comment (jayrhynas) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:352]: This could be written as:
```
case let err as LocalizedError where err.errorDescription != nil:
    return ": " + err.errorDescription
```

Comment (rjstelling) [reply Sources/ArgumentParser/Usage/UsageGenerator.swift]: I'm not sure why I have such a blind spot for switch/case patterns!

File: Sources/ArgumentParser/Parsable Properties/Errors.swift (hunk 5 above)
Comment (rjstelling) [reply Sources/ArgumentParser/Parsable Properties/Errors.swift]: @natecook1000 This change is no in a separate PR #119

File: Documentation/05 Validation and Errors.md [hunk 7]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
     }
 }
 ```
+
+## Handling Transform Errors
+
+During argument and option parsing you can use a closure to transform the command line stings to custom types. If this transformation fails you can throw a `ValidationError` its `message` property will be displayed to the user.
```
Comment (xwu) [inline Documentation/05 Validation and Errors.md:106]:
```suggestion
During argument and option parsing, you can use a closure to transform the command line strings to custom types. If this transformation fails, you can throw a `ValidationError`; its `message` property will be displayed to the user. 
```

File: Documentation/05 Validation and Errors.md [hunk 8]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
...
+  // a `ExampleDataModel` object using the JSONDecoder. If the
+  // string is not valid JSON `decode` will throw an error and
+  // parsing will halt.
+  @Argument(transform: {
+    guard let data = $0.data(using: .utf8) else { throw ValidationError("Badly encoded string, should be UTF-8") }
+    return try JSONDecoder().decode(ExampleDataModel.self, from: data) })
+  var inputJSON: ExampleDataModel
+  
+  // Specifiying this option will always cause the parser to exit
+  // and print the cistom error.
```
Comment (xwu) [inline Documentation/05 Validation and Errors.md:133]:
```suggestion
  // and print the custom error.
```

File: Documentation/05 Validation and Errors.md [hunk 9]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
...
+struct ExampleDataModel: Codable {
+  let identifier: UUID
+  let tokens: [String]
+  let tokenCount: Int
+}
+
+struct Example: ParsableCommand {
+
+  // Reads in the argument string and attempts to transform it to
+  // a `ExampleDataModel` object using the JSONDecoder. If the
```
Comment (xwu) [inline Documentation/05 Validation and Errors.md:124]:
```suggestion
  // a `ExampleDataModel` object using the `JSONDecoder`. If the
```

File: Documentation/05 Validation and Errors.md [hunk 10]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
...
+  // string is not valid JSON `decode` will throw an error and
```
Comment (xwu) [inline Documentation/05 Validation and Errors.md:125]:
```suggestion
  // string is not valid JSON, `decode` will throw an error and
```

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 11]
```diff
@@ -339,17 +339,33 @@ extension ErrorMessageGenerator {
     }
   }
   
-  func unableToParseValueMessage(origin: InputOrigin, name: Name?, value: String, key: InputKey) -> String {
+  func unableToParseValueMessage(origin: InputOrigin, name: Name?, value: String, key: InputKey, error: Error?) -> String {
     let valueName = arguments(for: key).first?.valueName
+    
+    // We want to make the "best effort" in producing a custom error message.
+    // We favour `LocalizedError.errorDescription` and fallback to
```
Comment (xwu) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:346]:
```suggestion
    // We favour `LocalizedError.errorDescription` and fall back to
```

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 12]
```diff
@@ -339,17 +339,33 @@ extension ErrorMessageGenerator {
...
+    // CustomStringConvertible. To opt-in return your custom error mesage
```
Comment (xwu) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:347]:
```suggestion
    // `CustomStringConvertible`. To opt in, return your custom error message
```

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 13]
```diff
@@ -339,17 +339,33 @@ extension ErrorMessageGenerator {
...
+    // in the `description` property of CustomStringConvertible.
```
Comment (xwu) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:348]:
```suggestion
    // as the `description` property of `CustomStringConvertible`.
```

File: Sources/ArgumentParser/Usage/UsageGenerator.swift [hunk 14]
```diff
@@ -339,17 +339,33 @@ extension ErrorMessageGenerator {
...
+    let customErrorMessage: String = {
+      switch error {
+        case let err as LocalizedError where err.errorDescription != nil:
```
Comment (xwu) [inline Sources/ArgumentParser/Usage/UsageGenerator.swift:351]:
```suggestion
      case let err as LocalizedError where err.errorDescription != nil:
```

**Comment by rjstelling** [comment]
@xwu Thanks! It must of been very late 😂

File: Documentation/02 Arguments, Options, and Flags.md [hunk 15]
```diff
@@ -178,7 +178,7 @@ struct Example: ParsableCommand {
 }
 ```
 
-Throw an error from the `transform` function to indicate that the user provided an invalid value for that type.
+Throw an error from the `transform` function to indicate that the user provided an invalid value for that type. (See: [Handling Transform Errors](./05%20Validation%20and%20Errors.md#handling-transform-errors) in [Validation and Errors](./05%20Validation%20and%20Errors.md) for details about customising `transform` function errors.)
```
Comment (natecook1000) [inline Documentation/02 Arguments, Options, and Flags.md:181]: Nits: This sentence doesn't need to be parenthesized, and let's use the American spelling here.
```suggestion
Throw an error from the `transform` function to indicate that the user provided an invalid value for that type. See [Handling Transform Errors](./05%20Validation%20and%20Errors.md#handling-transform-errors) for more about customizing `transform` function errors.
```

File: Documentation/05 Validation and Errors.md [hunk 16]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
...
+
+struct Example: ParsableCommand {
+
+  // Reads in the argument string and attempts to transform it to
+  // a `ExampleDataModel` object using the `JSONDecoder`. If the
+  // string is not valid JSON, `decode` will throw an error and
+  // parsing will halt.
+  @Argument(transform: {
+    guard let data = $0.data(using: .utf8) else { throw ValidationError("Badly encoded string, should be UTF-8") }
+    return try JSONDecoder().decode(ExampleDataModel.self, from: data) })
```
Comment (natecook1000) [inline Documentation/05 Validation and Errors.md:129]: This is a bit more code than I would like to encourage placing directly in the `@Argument` wrapper. Can you move it into the `ExampleDataModel`, either as an init or a static method that produces the type? Maybe `init(jsonString:) throws`, so that it doesn't match the `ExpressibleByArgument` initializer.

File: Documentation/05 Validation and Errors.md [hunk 17]
```diff
@@ -100,3 +100,38 @@ struct Example: ParsableCommand {
...
+  var inputJSON: ExampleDataModel
+  
+  // Specifiying this option will always cause the parser to exit
+  // and print the custom error.
+  @Option(transform: { throw ExampleTransformError(description: "Trying to write to failOption always produces an error. Input: \($0)") })
+  var failOption: String?
+}
+```
```
Comment (natecook1000) [inline Documentation/05 Validation and Errors.md:137]: Could you show an example of one or both of these errors in practice?

File: Tests/ArgumentParserEndToEndTests/TransformEndToEndTests.swift [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+fileprivate protocol Convert {
+    static func convert(_ str: String) throws -> Int
+}
+
+extension Convert {
+    
+    
+    static func convert(_ str: String) throws -> Int {
```
Comment (natecook1000) [inline Tests/ArgumentParserEndToEndTests/TransformEndToEndTests.swift:30]: Nit: Close this up, and convert to two space indents in this file.
```suggestion
    static func convert(_ str: String) throws -> Int {
```

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by rjstelling** [comment]
@natecook1000 Apologies, I requested a review and a commit had not been pushed. All corrected now.

**Comment by natecook1000** [comment]
@swift-ci Please test

**Comment by natecook1000** [comment]
@swift-ci Please test
